src/net/connection.rs

```diff
@@ -1,17 +1,129 @@
-mod connection_pool;
 mod quality;
-mod timeout_thread;
 mod virtual_connection;
 
-pub use self::connection_pool::ConnectionPool;
 pub use self::quality::{NetworkQuality, RttMeasurer};
-pub use self::timeout_thread::TimeoutThread;
 pub use self::virtual_connection::VirtualConnection;
 
-use std::collections::HashMap;
-use std::net::SocketAddr;
-use std::sync::{Arc, RwLock};
+use crate::config::Config;
+use std::{collections::HashMap, net::SocketAddr, time::Duration};
 
-pub type Connection = Arc<RwLock<VirtualConnection>>;
-pub type Connections = HashMap<SocketAddr, Connection>;
-pub type ConnectionsCollection = Arc<RwLock<Connections>>;
+/// Maintains a registry of active "connections". Essentially, when we receive a packet on the
+/// socket from a particular `SocketAddr`, we will track information about it here.
+pub struct ActiveConnections {
+    connections: HashMap<SocketAddr, VirtualConnection>,
+}
+
+impl ActiveConnections {
+    pub fn new() -> Self {
+        Self {
+            connections: HashMap::new(),
+        }
+    }
+
+    /// Try to get a VirtualConnection by address. If the connection does not exist, it will be
+    /// inserted and returned.
+    pub fn get_or_insert_connection(
```

Comment:
This is now blocking, correct? In that while we are adding a new connection, we can't process incoming data?

Reply:
its not blocking because there is no lock being acquired since no synchronization is required 😄

Reply:
With the async socket, there's only 1 thread reading, managing connections, and writing so technically speaking, when we are in this piece of code we are not actively reading from the socket but the OS is buffering up received data when we get to the reading phase of the loop.